pkg/perf/monitor.go

```diff
@@ -0,0 +1,105 @@
+package perf
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-co-op/gocron"
+	"github.com/gomodule/redigo/redis"
+
+	"github.com/pkg/errors"
+	"github.com/threefoldtech/zos/pkg/utils"
+)
+
+// PerformanceMonitor holds the module data
+type PerformanceMonitor struct {
+	scheduler *gocron.Scheduler
+	pool      *redis.Pool
+	tasks     []Task
+}
+
+// NewPerformanceMonitor returns PerformanceMonitor instance
+func NewPerformanceMonitor(redisAddr string) (*PerformanceMonitor, error) {
+	redisPool, err := utils.NewRedisPool(redisAddr)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed creating new redis pool")
+	}
+
+	scheduler := gocron.NewScheduler(time.UTC)
+
+	return &PerformanceMonitor{
+		scheduler: scheduler,
+		pool:      redisPool,
+		tasks:     []Task{},
+	}, nil
+}
+
+// AddTask a simple helper method to add new tasks
+func (pm *PerformanceMonitor) AddTask(task Task) {
+	pm.tasks = append(pm.tasks, task)
+}
+
+// runTask runs the task and store its result
+func (pm *PerformanceMonitor) runTask(ctx context.Context, task Task) error {
+	res, err := task.Run(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to run task: %s", task.ID())
+	}
+
+	err = pm.setCache(ctx, TaskResult{
+		Name:      task.ID(),
+		Timestamp: uint64(time.Now().Unix()),
+		Result:    res,
+	})
+	if err != nil {
+		return errors.Wrap(err, "failed to set cache")
+	}
+
+	return nil
+}
+
+// Run adds the tasks to the corn queue and start the scheduler
+func (pm *PerformanceMonitor) Run(ctx context.Context) error {
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(pm.tasks))
+
+	for _, task := range pm.tasks {
+		_, err := pm.scheduler.CronWithSeconds(task.Cron()).Do(func() error {
+			return pm.runTask(ctx, task)
+		})
+		if err != nil {
+			return errors.Wrapf(err, "failed to schedule the task: %s", task.ID())
+		}
+
+		wg.Add(1)
+		go func(task Task) {
+			var routineErr error
+			defer wg.Done()
+			ok, err := pm.exists(task.ID())
+			if err != nil {
+				routineErr = errors.Wrapf(err, "failed to find key %s", task.ID())
+			}
+
+			if !ok {
+				if err := pm.runTask(ctx, task); err != nil {
+					routineErr = errors.Wrapf(err, "failed to run task: %s", task.ID())
+				}
+			}
+			errChan <- routineErr
+		}(task)
+
+	}
+
+	wg.Wait()
+	close(errChan)
```

Comment:
I am wondering actually if we need to `wait` until the tests are done ? why do they need to be handled any different than when the `cron` starts them.

This will slow down the start of the main service (blocking on Run()) for a very long time on first run. What i wanted to do is that if there are no cached results u `start` the tests in the background to have some results, and then leave it to the shceulder for later runs. Very simple, there should no be any blocking in Run in this case, if an perf test failed u should only see a log in the daemon logs but that's it.